### PR TITLE
virtio_serial: Add test case for port and pci hotplug

### PR DESCRIPTION
**Pull request by vivianQizhu**
id: 1744411
Signed-off-by: Qianqian Zhu <qizhu@redhat.com>

### CODE REVIEW

File: qemu/tests/virtio_console.py [hunk 1]
```diff
@@ -37,6 +39,40 @@ def __set_exit_event():
...
+    devices = []
+    for index, chardev in enumerate(char_devices):
+        chardev_param = params.object_params(chardev)
+        file_name = vm.get_serial_console_filename(chardev)
+        backend = chardev_param.get('chardev_backend',
+                                    'unix_socket')
+        if backend in ['udp', 'tcp_socket']:
+            chardev_param['chardev_host'] = host
+            chardev_param['chardev_port'] = free_ports[index]
+        device = devices.chardev_define_by_params(chardev, chardev_param, file_name)
```
Comment (nanliu-r) [inline qemu/tests/virtio_console.py]: Test case:
(1/1) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.x86_64.io-github-autotest-qemu.virtio_port_hotplug.hotplug_port_pci: ERROR: 'list' object has no attribute 'chardev_define_by_params' (3.46 s)

02:05:23 ERROR| File "/home/kar/kar/workspace/var/lib/avocado/data/avocado-vt/test-providers.d/downloads/io-github-autotest-qemu/qemu/tests/virtio_serial_hotplug_port_pci.py", line 18, in run
02:05:23 ERROR| char_devices = add_unused_chardev_cmdline(vm, params)
02:05:23 ERROR| File "/home/kar/kar/workspace/var/lib/avocado/data/avocado-vt/test-providers.d/downloads/io-github-autotest-qemu/qemu/tests/virtio_console.py", line 69, in add_unused_chardev_cmdline
02:05:23 ERROR| device = devices.chardev_define_by_params(chardev, chardev_param, file_name)
02:05:23 ERROR| AttributeError: 'list' object has no attribute 'chardev_define_by_params

Comment (vivianQizhu) [reply qemu/tests/virtio_console.py]: @nanliu-r Updated, thanks.

**Comment by nanliu-r** [comment]
Test PASS and Command line looks good.
 (1/1) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.x86_64.io-github-autotest-qemu.virtio_port_hotplug.hotplug_port_pci.q35: PASS (47.19 s)

Command line:
    -chardev socket,server,id=channel1,nowait,path=/var/tmp/avocado_8z2yieoa/serial-channel1-20190821-224114-wdYCv9oe \
    -chardev socket,server,id=channel2,nowait,path=/var/tmp/avocado_8z2yieoa/serial-channel2-20190821-224114-wdYCv9oe \
    -enable-kvm \
    -device pcie-root-port,id=pcie_extra_root_port_0,slot=5,chassis=5,addr=0x5,bus=pcie.0

**Comment by vivianQizhu** [comment]
PLEASE DO NOT MERGE IT since not completed now, post in advance just for sharing common usage.

**Comment by vivianQizhu** [comment]
Need test together with https://github.com/avocado-framework/avocado-vt/pull/1982

File: qemu/tests/virtio_console.py [hunk 2]
```diff
@@ -37,6 +39,81 @@ def __set_exit_event():
...
+        5000, 6000, len(char_devices), host)
+    device_list = []
+    for index, chardev in enumerate(char_devices):
+        chardev_param = params.object_params(chardev)
+        file_name = vm.get_serial_console_filename(chardev)
+        backend = chardev_param.get('chardev_backend',
+                                    'unix_socket')
+        if backend in ['udp', 'tcp_socket']:
+            chardev_param['chardev_host'] = host
+            chardev_param['chardev_port'] = free_ports[index]
```
Comment (nanliu-r) [inline qemu/tests/virtio_console.py]: ‘chardev_port’ type needs to be a string while doing hotplug.

07:23:04 ERROR|   File "/home/kar/kar/workspace/var/lib/avocado/data/avocado-vt/test-providers.d/downloads/io-github-autotest-qemu/qemu/tests/virtio_serial_hotplug_port_various_chardev.py", line 39, in run
07:23:04 ERROR|     vm.devices.simple_hotplug(chardevice, vm.monitor)
07:23:04 ERROR|   File "/home/kar/kar/workspace/avocado-vt/virttest/qemu_devices/qcontainer.py", line 678, in simple_hotplug
07:23:04 ERROR|     out = device.hotplug(monitor)
07:23:04 ERROR|   File "/home/kar/kar/workspace/avocado-vt/virttest/qemu_devices/qdevices.py", line 257, in hotplug
07:23:04 ERROR|     return monitor.cmd(cmd, args)
07:23:04 ERROR|   File "/home/kar/kar/workspace/avocado-vt/virttest/qemu_monitor.py", line 1943, in cmd
07:23:04 ERROR|     raise QMPCmdError(cmd, args, r["error"])
07:23:04 ERROR| virttest.qemu_monitor.QMPCmdError: QMP command 'chardev-add' failed    (arguments: {'id': 'channel2', 'backend': {'type': 'socket', 'data': {'addr': {'type': 'inet', 'data': {'host': '127.0.0.1', 'port': 5142}}}}},    error message: {'class': 'GenericError', 'desc': "Invalid parameter type for 'backend.data.addr.data.port', expected: string"}

Comment (nanliu-r) [inline qemu/tests/virtio_console.py]: port type needs to be a string while doing hotplug a tcp_socket chardev.

07:23:04 DEBUG| Send command: {'execute': 'chardev-add', 'arguments': {'id': 'channel2', 'backend': {'type': 'socket', 'data': {'addr': {'type': 'inet', 'data': {'host': '127.0.0.1', 'port': 5142}}}}}, 'id': 'lTrq4LVk'}
......
07:23:04 ERROR|   File "/home/kar/kar/workspace/var/lib/avocado/data/avocado-vt/test-providers.d/downloads/io-github-
07:23:04 ERROR| virttest.qemu_monitor.QMPCmdError: QMP command 'chardev-add' failed    (arguments: {'id': 'channel2', 'backend': {'type': 'socket', 'data': {'addr': {'type': 'inet', 'data': {'host': '127.0.0.1', 'port': 5142}}}}},    error message: {'class': 'GenericError', 'desc': "Invalid parameter type for 'backend.data.addr.data.port', expected: string"})

Comment (vivianQizhu) [reply qemu/tests/virtio_console.py]: @nanliu-r Updated to str, thanks.

**Comment by nanliu-r** [comment]
Test pass and steps are reasonable, thanks.

 (4/4) repeat4.Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.x86_64.io-github-autotest-qemu.virtio_port_hotplug.hotplug_port_pci: PASS (65.96 s)
 (4/4) repeat4.Host_RHEL.m7.u7.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.7.x86_64.io-github-autotest-qemu.virtio_port_hotplug.hotplug_port_pci.q35: PASS (142.85 s)

repeat1.Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.virtio_port_hotplug.hotplug_port_pci: PASS (130.44 s)

**Comment by nanliu-r** [comment]
ACK.

**Pull request merged**